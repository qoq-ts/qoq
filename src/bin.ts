#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import chalk from 'chalk';
import { exec } from 'child_process';
import { hideBin } from 'yargs/helpers';

const isESM = typeof require === 'undefined';
const files = ['./src/console', './lib/console', './console'];

function js(ext: '.js' | '.cjs' | '.mjs') {
  let jsFile: string | undefined;

  for (let i = 0; i < files.length; ++i) {
    const file = path.resolve(files[i] + ext);

    if (fs.existsSync(file)) {
      jsFile = file;
      break;
    }
  }

  if (jsFile) {
    import(jsFile);
    return true;
  }

  return false;
}

function ts() {
  let tsFile: string | undefined;

  for (let i = 0; i < files.length; ++i) {
    const file = path.resolve(files[i] + '.ts');

    if (fs.existsSync(file)) {
      tsFile = file;
      break;
    }
  }

  if (tsFile) {
    // User should install ts-node manually
    if (isESM) {
      const node = process.argv[0];
      const args = hideBin(process.argv);
      const nodeOptions = [
        '--no-warnings',
        '--loader=ts-node/esm/transpile-only',
        '--experimental-specifier-resolution=node',
      ];
      tsFile = tsFile!.replace(/\.ts$/, '.js');

      exec(
        `${node} ${nodeOptions.join(' ')} ${tsFile} ${args}`,
        {
          cwd: process.cwd(),
        },
        (err, stdout) => {
          if (err) {
            throw err;
          }

          process.stdout.write(stdout);
        },
      );
    } else {
      import('ts-node/register/transpile-only').then(() => {
        import(tsFile!);
      });
    }

    return true;
  }

  return false;
}

if (!js('.js') && !js(isESM ? '.mjs' : '.cjs') && !ts()) {
  console.error(chalk.red('Command entry file console.{ts|js|mjs|cjs} is not found.'));
  process.exit(127);
}
