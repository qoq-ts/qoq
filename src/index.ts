// Basic
export { WebRouter } from './router/WebRouter';
export { ConsoleRouter } from './router/ConsoleRouter';
export { WebSlotManager, ConsoleSlotManager } from './slot/SlotManager';
export { Slot } from './slot/Slot';
export { Cache } from './slot/Cache';
export { version } from './util/version';
export { createConfig } from './util/createConfig';
export { validator } from './validator';
export { WebApplication } from './core/WebApplication';
export { ConsoleApplication } from './core/ConsoleApplication';
export { FileCacheOptions, FileCache } from './caching/FileCache';
export { MemoryCacheOptions, MemoryCache } from './caching/MemoryCache';

// Advanced
export { compose, Composer } from './util/compose';
export { Method } from './util/Method';
export { Next, ConsoleSlotCtx, WebSlotCtx, MixSlotCtx } from './slot/Slot';
export { ValidatorStatic } from './validator';
export { Validator, ValidatorOptions, ValidatorType } from './validator/Validator';
export { WebCtx } from './core/WebContext';
export { ConsoleCtx } from './core/ConsoleContext';
export { HttpError } from 'http-errors';
export { BaseCache, BaseCacheOptions } from './caching/BaseCache';
