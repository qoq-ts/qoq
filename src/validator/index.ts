import { NumberValidator } from './NumberValidator';
import { StringValidator } from './StringValidator';
import { EnumValidator } from './EnumValidator';
import { UrlValidator } from './UrlValidator';
import { ArrayValidator } from './ArrayValidator';
import { JsonValidator } from './JsonValidator';
import { BooleanValidator } from './BooleanValidator';
import { FileValidator } from './FileValidator';
import { EmailValidator } from './EmailValidator';
import { UUIDValidator } from './UUIDValidator';
import { validate } from './validate';
import { IPValidator } from './IPValidator';
import { TimestampValidator } from './TimestampValidator';

export class ValidatorStatic {
  readonly validate = validate;

  get number(): NumberValidator {
    return new NumberValidator();
  }

  get integer(): NumberValidator {
    return new NumberValidator().onlyInteger(true);
  }

  get enum(): EnumValidator {
    return new EnumValidator();
  }

  get string(): StringValidator {
    return new StringValidator();
  }

  get url(): UrlValidator {
    return new UrlValidator();
  }

  get array(): ArrayValidator {
    return new ArrayValidator();
  }

  get json(): JsonValidator {
    return new JsonValidator();
  }

  get boolean(): BooleanValidator {
    return new BooleanValidator();
  }

  get file(): FileValidator {
    return new FileValidator();
  }

  get email(): EmailValidator {
    return new EmailValidator();
  }

  get uuid(): UUIDValidator {
    return new UUIDValidator;
  }

  get ip(): IPValidator {
    return new IPValidator();
  }

  get timestamp(): TimestampValidator {
    return new TimestampValidator();
  }
}

export const validator = new ValidatorStatic();
