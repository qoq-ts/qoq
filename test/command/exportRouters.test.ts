import { existsSync, readFileSync, unlinkSync } from 'fs';
import { tmpdir } from 'os';
import { ConsoleApplication } from '../../src';

const app = new ConsoleApplication({
  commandsDir: './src/command',
});

const input = './test/fixture/router-schema';

it ('generate snapshot', async () => {
  try {
    unlinkSync(input + '/snapshot-formatted.json');
  } catch {}
  expect(existsSync(input + '/snapshot-formatted.json')).toBeFalsy();

  await app.execute(
    'export:routers',
    '-i', input,
    '-o', input + '/snapshot-formatted.json',
    '-f'
  );

  expect(existsSync(input + '/snapshot-formatted.json')).toBeTruthy();
});

it ('can export web routers to file', async () => {
  const output = tmpdir() + '/snapshot.json';

  await app.execute(
    'export:routers',
    '-i', input,
    '-o', output,
  );

  expect(readFileSync(output).toString()).toMatchSnapshot();
});


it ('can export web routers to context', async () => {
  const output = tmpdir() + '/snapshot.json';

  const ctx = await app.execute(
    'export:routers',
    '-i', input,
    '-o', output,
  );

  // @ts-expect-error
  const data = JSON.stringify(ctx.state.routers);
  expect(data).toMatchSnapshot();
});

it ('can format output json', async () => {
  const output = tmpdir() + '/snapshot.json';

  await app.execute(
    'export:routers',
    '-i', input,
    '-o', output,
    '-f'
  );

  expect(readFileSync(output).toString()).toMatchSnapshot();
});
