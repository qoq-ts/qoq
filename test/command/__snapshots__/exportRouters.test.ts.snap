// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can export web routers to context 1`] = `"[{\\"uris\\":[\\"/projects\\"],\\"method\\":\\"GET\\",\\"title\\":\\"Get Projects\\",\\"description\\":\\"It is a desc test\\",\\"response\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"object\\",\\"validator\\":\\"json\\",\\"properties\\":{\\"page\\":{\\"label\\":\\"\\",\\"description\\":\\"Current Page\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"result\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"array\\",\\"validator\\":\\"array\\",\\"items\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"object\\",\\"validator\\":\\"json\\",\\"properties\\":{\\"id\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}}}},\\"total\\":{\\"label\\":\\"\\",\\"description\\":\\"Total records amount\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"}}},\\"headers\\":{\\"x-date\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"},\\"y-schema\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"number\\",\\"validator\\":\\"number\\"}},\\"query\\":{\\"page\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"pageSize\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":10,\\"required\\":false,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"Project Name\\",\\"description\\":\\"You can filter the project name\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}},{\\"uris\\":[\\"/p/users\\"],\\"method\\":\\"GET\\",\\"title\\":\\"Get Users\\",\\"description\\":\\"Get all users from db\\",\\"response\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"url\\"},\\"headers\\":{},\\"query\\":{\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"You can filter the project name\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}},{\\"uris\\":[\\"/projects\\"],\\"method\\":\\"POST\\",\\"title\\":\\"Create a project\\",\\"description\\":\\"Create a project description\\",\\"response\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"object\\",\\"validator\\":\\"json\\",\\"properties\\":{\\"id\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"}}},\\"headers\\":{},\\"query\\":{\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"Project search name\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"},\\"description\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}},{\\"uris\\":[\\"/p/projects\\"],\\"method\\":\\"POST\\",\\"title\\":\\"\\",\\"description\\":\\"\\",\\"headers\\":{},\\"query\\":{\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"Project search name\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"},\\"description\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}}]"`;

exports[`can export web routers to file 1`] = `"[{\\"uris\\":[\\"/projects\\"],\\"method\\":\\"GET\\",\\"title\\":\\"Get Projects\\",\\"description\\":\\"It is a desc test\\",\\"response\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"object\\",\\"validator\\":\\"json\\",\\"properties\\":{\\"page\\":{\\"label\\":\\"\\",\\"description\\":\\"Current Page\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"result\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"array\\",\\"validator\\":\\"array\\",\\"items\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"object\\",\\"validator\\":\\"json\\",\\"properties\\":{\\"id\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}}}},\\"total\\":{\\"label\\":\\"\\",\\"description\\":\\"Total records amount\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"}}},\\"headers\\":{\\"x-date\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"},\\"y-schema\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"number\\",\\"validator\\":\\"number\\"}},\\"query\\":{\\"page\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"pageSize\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":10,\\"required\\":false,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"Project Name\\",\\"description\\":\\"You can filter the project name\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}},{\\"uris\\":[\\"/p/users\\"],\\"method\\":\\"GET\\",\\"title\\":\\"Get Users\\",\\"description\\":\\"Get all users from db\\",\\"response\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"url\\"},\\"headers\\":{},\\"query\\":{\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"You can filter the project name\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}},{\\"uris\\":[\\"/projects\\"],\\"method\\":\\"POST\\",\\"title\\":\\"Create a project\\",\\"description\\":\\"Create a project description\\",\\"response\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"object\\",\\"validator\\":\\"json\\",\\"properties\\":{\\"id\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"}}},\\"headers\\":{},\\"query\\":{\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"Project search name\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"},\\"description\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}},{\\"uris\\":[\\"/p/projects\\"],\\"method\\":\\"POST\\",\\"title\\":\\"\\",\\"description\\":\\"\\",\\"headers\\":{},\\"query\\":{\\"projectId\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"integer\\",\\"validator\\":\\"number\\"},\\"name\\":{\\"label\\":\\"Project search name\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":true,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"},\\"description\\":{\\"label\\":\\"\\",\\"description\\":\\"\\",\\"defaultValue\\":\\"\\",\\"required\\":false,\\"type\\":\\"string\\",\\"validator\\":\\"string\\"}},\\"body\\":{},\\"params\\":{},\\"additional\\":{}}]"`;

exports[`can format output json 1`] = `
"[
    {
        \\"uris\\": [
            \\"/projects\\"
        ],
        \\"method\\": \\"GET\\",
        \\"title\\": \\"Get Projects\\",
        \\"description\\": \\"It is a desc test\\",
        \\"response\\": {
            \\"label\\": \\"\\",
            \\"description\\": \\"\\",
            \\"defaultValue\\": \\"\\",
            \\"required\\": true,
            \\"type\\": \\"object\\",
            \\"validator\\": \\"json\\",
            \\"properties\\": {
                \\"page\\": {
                    \\"label\\": \\"\\",
                    \\"description\\": \\"Current Page\\",
                    \\"defaultValue\\": \\"\\",
                    \\"required\\": true,
                    \\"type\\": \\"integer\\",
                    \\"validator\\": \\"number\\"
                },
                \\"result\\": {
                    \\"label\\": \\"\\",
                    \\"description\\": \\"\\",
                    \\"defaultValue\\": \\"\\",
                    \\"required\\": true,
                    \\"type\\": \\"array\\",
                    \\"validator\\": \\"array\\",
                    \\"items\\": {
                        \\"label\\": \\"\\",
                        \\"description\\": \\"\\",
                        \\"defaultValue\\": \\"\\",
                        \\"required\\": true,
                        \\"type\\": \\"object\\",
                        \\"validator\\": \\"json\\",
                        \\"properties\\": {
                            \\"id\\": {
                                \\"label\\": \\"\\",
                                \\"description\\": \\"\\",
                                \\"defaultValue\\": \\"\\",
                                \\"required\\": true,
                                \\"type\\": \\"integer\\",
                                \\"validator\\": \\"number\\"
                            },
                            \\"name\\": {
                                \\"label\\": \\"\\",
                                \\"description\\": \\"\\",
                                \\"defaultValue\\": \\"\\",
                                \\"required\\": true,
                                \\"type\\": \\"string\\",
                                \\"validator\\": \\"string\\"
                            }
                        }
                    }
                },
                \\"total\\": {
                    \\"label\\": \\"\\",
                    \\"description\\": \\"Total records amount\\",
                    \\"defaultValue\\": \\"\\",
                    \\"required\\": true,
                    \\"type\\": \\"integer\\",
                    \\"validator\\": \\"number\\"
                }
            }
        },
        \\"headers\\": {
            \\"x-date\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            },
            \\"y-schema\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": false,
                \\"type\\": \\"number\\",
                \\"validator\\": \\"number\\"
            }
        },
        \\"query\\": {
            \\"page\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"integer\\",
                \\"validator\\": \\"number\\"
            },
            \\"pageSize\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": 10,
                \\"required\\": false,
                \\"type\\": \\"integer\\",
                \\"validator\\": \\"number\\"
            },
            \\"projectId\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"integer\\",
                \\"validator\\": \\"number\\"
            },
            \\"name\\": {
                \\"label\\": \\"Project Name\\",
                \\"description\\": \\"You can filter the project name\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": false,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            }
        },
        \\"body\\": {},
        \\"params\\": {},
        \\"additional\\": {}
    },
    {
        \\"uris\\": [
            \\"/p/users\\"
        ],
        \\"method\\": \\"GET\\",
        \\"title\\": \\"Get Users\\",
        \\"description\\": \\"Get all users from db\\",
        \\"response\\": {
            \\"label\\": \\"\\",
            \\"description\\": \\"\\",
            \\"defaultValue\\": \\"\\",
            \\"required\\": true,
            \\"type\\": \\"string\\",
            \\"validator\\": \\"url\\"
        },
        \\"headers\\": {},
        \\"query\\": {
            \\"projectId\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"integer\\",
                \\"validator\\": \\"number\\"
            },
            \\"name\\": {
                \\"label\\": \\"You can filter the project name\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": false,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            }
        },
        \\"body\\": {},
        \\"params\\": {},
        \\"additional\\": {}
    },
    {
        \\"uris\\": [
            \\"/projects\\"
        ],
        \\"method\\": \\"POST\\",
        \\"title\\": \\"Create a project\\",
        \\"description\\": \\"Create a project description\\",
        \\"response\\": {
            \\"label\\": \\"\\",
            \\"description\\": \\"\\",
            \\"defaultValue\\": \\"\\",
            \\"required\\": true,
            \\"type\\": \\"object\\",
            \\"validator\\": \\"json\\",
            \\"properties\\": {
                \\"id\\": {
                    \\"label\\": \\"\\",
                    \\"description\\": \\"\\",
                    \\"defaultValue\\": \\"\\",
                    \\"required\\": true,
                    \\"type\\": \\"integer\\",
                    \\"validator\\": \\"number\\"
                }
            }
        },
        \\"headers\\": {},
        \\"query\\": {
            \\"projectId\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"integer\\",
                \\"validator\\": \\"number\\"
            },
            \\"name\\": {
                \\"label\\": \\"Project search name\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            },
            \\"description\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": false,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            }
        },
        \\"body\\": {},
        \\"params\\": {},
        \\"additional\\": {}
    },
    {
        \\"uris\\": [
            \\"/p/projects\\"
        ],
        \\"method\\": \\"POST\\",
        \\"title\\": \\"\\",
        \\"description\\": \\"\\",
        \\"headers\\": {},
        \\"query\\": {
            \\"projectId\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"integer\\",
                \\"validator\\": \\"number\\"
            },
            \\"name\\": {
                \\"label\\": \\"Project search name\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": true,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            },
            \\"description\\": {
                \\"label\\": \\"\\",
                \\"description\\": \\"\\",
                \\"defaultValue\\": \\"\\",
                \\"required\\": false,
                \\"type\\": \\"string\\",
                \\"validator\\": \\"string\\"
            }
        },
        \\"body\\": {},
        \\"params\\": {},
        \\"additional\\": {}
    }
]"
`;
